src/clj/clojure/next.clj

```diff
@@ -792,18 +797,89 @@
   (platform-flush)
   nil)
 
-(defn print-simple [o w]
-  (print-meta o w)
-  (platform-write w (str o)))
-
-(defmulti print-method (fn [x wrtr]
-                         (let [t (get (meta x) :type)]
-                           (if (keyword? t) t (class x)))))
+(defn print-ctor [obj print-args wrtr]
+  (platform-print-constructor obj print-args wrtr))
+
+(defn- print-meta [o w]
+  (when-let [m (meta o)]
+    (when (and (pos? (count m))
+            (or *print-dup*
+              (and *print-meta* *print-readably*)))
+      (platform-write w "^")
+      (if (and (= (count m) 1) (:tag m))
+          (print-method (:tag m) w)
+          (print-method m w))
+      (platform-write w " "))))
+
+(defn print-simple [obj wrtr]
+  (print-meta obj wrtr)
+  (platform-write wrtr (str obj)))
+
+(defn- print-sequential [begin, print-one, sep, end, sequence, w]
+  (binding [*print-level* (and (not *print-dup*) *print-level* (dec *print-level*))]
+    (if (and *print-level* (neg? *print-level*))
+      (platform-write w "#")
+      (do
+        (platform-write w begin)
+        (when-let [xs (seq sequence)]
+          (if (and (not *print-dup*) *print-length*)
+            ; TODO: switch to [[x & xs] xs] once
+            ; it is using clojure.core/nth
+            (loop [rxs xs
+                   print-length *print-length*]
+              (let [x (first rxs)
+                    xs (next rxs)]
+                (if (zero? print-length)
+                  (platform-write w "...")
+                  (do
+                    (print-one x w)
+                    (when xs
+                      (platform-write w sep)
+                      (recur xs (dec print-length)))))))
+            ; TODO: switch to [[x & xs] xs] once
+            ; it is using clojure.core/nth
+            (loop [rxs xs]
+              (let [x (first rxs)
+                    xs (next rxs)]
+                (print-one x w)
+                (when xs
+                  (platform-write w sep)
+                  (recur xs))))))
+        (platform-write w end)))))
```

Comment:
This method is huge. I understand why, but I'd like to see a lot more tests around this behavior. I count at least 6 `if` statements and only two tests.